include/common/oclengine.hpp

```diff
@@ -107,14 +107,18 @@ class OCLEngine {
 public:
     /// Get a pointer to the Instance of the singleton. (The instance will be instantiated, if it does not exist yet.)
     static OCLEngine* Instance();
-    /// Get a pointer to the OpenCL context
+    /// Get a pointer one of the available OpenCL contexts, by its index in the list of all contexts.
     DeviceContextPtr GetDeviceContextPtr(const int& dev = -1);
-    int GetDeviceCount() { return deviceCount; }
+    /// Get the list of all available devices (and their supporting objects).
+    std::vector<DeviceContextPtr> GetDeviceContextPtrVector();
+    /// Set the list of DeviceContextPtr object available for use. If one takes the result of GetDeviceContextPtrVector(), trims items from it, and sets it with this method, (at initialization, before any QEngine objects depend on them,) all resources associated with the removed items are freed.
+    void SetDeviceContextPtrVector(std::vector<DeviceContextPtr> vec, DeviceContextPtr dcp = nullptr);
```

Comment:
Do you want a "full rebuild" option to repopulate the device list?

Reply:
I think it's more likely to encourage bad behavior than good. The available devices won't change on the fly. Based on VirtualCL docs, even in a clustered environment, that's the expectation. You kill and restart the executable. However, should there be demand, we just make the initialization method public, and that will do it.

Reply:
To clarify, you can't safely rebuild the list unless all QEngineOCL and QEngineOCLMulti instances are killed, so they give up their OCLDeviceContext instances. So, it's likely to encourage bad behavior.